include/zio/zio_buf.h

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (c) 2019 Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file zio_buf.h
+ *
+ * @brief ZIO Buffer provides a pollable interface to a FIFO like buffer
+ *
+ * Device drivers implement this interface using either a software or
+ * hardware FIFO.
+ *
+ * The format of the buffer is device dependent and thus a device must
+ * implement part of the API by describing the layout of the buffer. Each
+ * device is responsible for self describing the position, byte size, byte
+ * order, bit precision, and bit shift each value provides in a static
+ * array of struct zio_chan values.
+ *
+ * Changing enabled/disabled channels while using the same zio_buf isn't
+ * supported for that reason.
+ *
+ * In addition to the data itself a 32 bit trigger timestamp corresponding to
+ * k_cycle_get_32() may be placed as the last set of bytes of each datum
+ * pulled.
+ *
+ * The buffer is pollable using k_poll where the pollable event becomes
+ * available when the zio_buf contains at least the designated watermark
+ * number of datums. It will cease to become pollable once it contains
+ * less than the designated watermark number of datums. 
+ *
+ * Adjusting the watermark value may or may not be implemented depending
+ * on the FIFO implementation.
+ *
+ * The ability to poll one or more buffers allows an application designer
+ * to decide how to deal with incoming streams of data from one or more
+ * sensors.
+ *
+ * This is heavily inspired by the fantastic work in Linux's IIO subsystem,
+ * with a microcontroller focused RTOS in mind.
+ *
+ * As an example in the case of a 9 DoF IMU sensor where each channel
+ * produces a 16 bit value the buffer after each trigger without timestamps
+ * would contain 18 bytes of data, 22 bytes with the 32bit cycle counter
+ * timestamp.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_BUF_H_
+#define ZEPHYR_INCLUDE_ZIO_BUF_H_
+
+#include <zephyr/types.h>
+#include <kernel.h>
+
+#include <zio/zio_dev.h>
+
+struct zio_buf;
+
+/**
+ * @brief Function to pull from the zio_buf a single datum
+ *
+ * The size of the datum is decided by the device and its set of active
+ * channels. Its expected the caller already knows the exact size of
+ * sample.
+ */
+typedef int (*zio_buf_pull_t)(struct zio_buf *buf, void *datum);
+
+/**
+ * @brief Function to setup a poll event for the zio_buf
+ *
+ * The device itself decides when data becomes available and how. Then in
+ * a zio application a k_poll_event is setup as a way of polling one or
+ * more buffers.
+ */
+typedef int (*zio_buf_poll_init_t)(struct zio_buf *buf,
```

Comment:
This is reversal of current k_poll() usage. Here, you are passing the k_poll_event to the ZIO in order to initialize it while in all other cases you are passing an object to the k_poll_event initializer. I would suggest keeping consistent usage patterns on the API.

Also, I do not see k_poll() update which would create possibility to poll on this buffer.

Reply:
Right, part of this was my indecision about whether I should expose the pollable type in the public zio_buf struct or leave it hidden behind a implementable function pointer for each implementor.

Reply:
I would put this in zio_buf. Then the zio_buf could call implementation and later signal that new data arrived.

Reply:
That I think makes sense, I'd really like to make zio_buf pollable itself like you mentioned in your comments, it wasn't clear to me how to best go about that.

Reply:
Making a new poll-able object is pretty simple. Look for  z_handle_obj_poll_events() and _POLL_EVENT in the code.

Reply:
I started working on doing this in another branch, I like where its taking this patch quite a lot!

Reply:
Please use a semaphore instead of adding a new type of polling object

Reply:
I've gone ahead and used a semaphore given @carlescufi and @andrewboie suggested doing so, adding another pollable type turns out to touch a lot more than I initially expected (needs changed in kernel/poll.c to work)

Reply:
@andrewboie: Why? If we make the zio buffer a pollable object, it will be smaller as well as easier to use.